### PR TITLE
V 1.0 Major Update

### PR DESCRIPTION
**Pull request by liruilong940607**
This is a continue development based on #172. There are many major features in this update, including:

- Comparing to the [official implementation](https://github.com/graphdeco-inria/gaussian-splatting), gsplat enables up to **4x less training memory footprint**, and up to **2x less training time** on Mip-NeRF 360 captures, and potential more on larger scenes. 

- Support extremely large scene rendering, which is magnitudes faster than the official CUDA backend [diff-gaussian-rasterization](https://github.com/graphdeco-inria/diff-gaussian-rasterization). 

- Extra features, including batch rasterization, N-D feature rendering (faster), depth rendering, sparse gradient etc.

### CODE REVIEW

File: gsplat/__init__.py
Comment (maturk) [inline gsplat/__init__.py:5]: the version should probably be bumped from 0.1.11 to 1.0.0

File: tests/test_basic.py
Comment (maturk) [inline tests/test_basic.py:1]: I have a few tests failing on my machine:

image

**Review by maturk — COMMENTED** [review]
Fantastic stuff, I left some small comments but will look at this more in detail.

**Comment by jb-ye** [comment]
Let's have a one version update in 0.1.xx to check in those untagged commits before we bump into 1.0

File: docs/source/migration/migration_inria.rst
Comment (liruilong940607) [reply docs/source/migration/migration_inria.rst:1]: done

File: gsplat/__init__.py
Comment (liruilong940607) [reply gsplat/__init__.py:20]: done

File: tests/test_basic.py
Comment (liruilong940607) [reply tests/test_basic.py:1]: I wonder if you can reproduces it. I also experienced some numerical instability on my end.

**Comment by liruilong940607** [comment]
Ok I think I've cleanup all the TODOs in my plates. I'm a bit afraid of merging it into the main with such a big update. Anyone wants to mess around with it or proof read the doc?  @jb-ye @maturk @brentyi @jake-austin @vye16  

Here is the preview version of the doc as for the latest commit: https://6656ab91e5ef673fdf2219e3--inquisitive-mandazi-513b7e.netlify.app/

I'll pin the current main to a 0.1.x version before merging this PR.

**Comment by liruilong940607** [comment]
Ok seems like I missed the cleanup on .md files, and also needs to fix the testing is somehow failing in github actions

**Comment by jb-ye** [comment]
> With that, I think I'm fine with not blocking the v1.0 release by these extra supports

Sounds good to me.

**Comment by liruilong940607** [comment]
Ok fixed the tests and update the md files. So good to go now.

**Comment by liruilong940607** [comment]
Just pinned main to `0.1.12`

File: gsplat/cuda/_wrapper.py
Comment (hturki) [inline gsplat/cuda/_wrapper.py:143]: (noted this when I was porting https://github.com/nerfstudio-project/gsplat/compare/cuda_v2...hturki:gsplat:ht/cuda_v2_batch?expand=1 to this branch)

I would suggest annotating the parameters that might be None as Optional[Tensor]

File: gsplat/cuda/csrc/rasterization.cu
Comment (vye16) [inline gsplat/cuda/csrc/rasterization.cu:79]: nice!

File: gsplat/cuda/_wrapper.py
Comment (vye16) [inline gsplat/cuda/_wrapper.py:141]: projection = quat_scale_to_covar -> world_to_cam -> persp_proj
maybe rename to something like full_gaussian_projection so it's not too similar with persp_proj or something?

File: gsplat/cuda/_wrapper.py
Comment (vye16) [inline gsplat/cuda/_wrapper.py:666]: same comment about name

File: gsplat/cuda/csrc/rasterization.cu
Comment (vye16) [inline gsplat/cuda/csrc/rasterization.cu:107]: nit: totel_elems -> total_elems

File: gsplat/cuda/_wrapper.py
Comment (liruilong940607) [reply gsplat/cuda/_wrapper.py:143]: oh yes good catch! I think was generally following this rule but I clearly missed this one!

File: gsplat/cuda/csrc/rasterization.cu
Comment (vye16) [inline gsplat/cuda/csrc/rasterization.cu:254]: okay so this function only rasterizes a range of indices determined by step0 to step1
maybe rename to rasterize_to_indices_in_range_kernel and rename step0 and step1 to range_start and range_end respectively?

File: gsplat/cuda/csrc/rasterization.cu
Comment (vye16) [inline gsplat/cuda/csrc/rasterization.cu:478]: really nice, also parallelize of cameras with multidim thread block 👍 👍

File: gsplat/cuda/csrc/rasterization.cu
Comment (vye16) [inline gsplat/cuda/csrc/rasterization.cu:678]: good idea, though not necessary for this version unless you're motivated XD

File: gsplat/cuda/_wrapper.py
Comment (vye16) [inline gsplat/cuda/_wrapper.py:201]: I think this needs to be renamed to be clearer, row/column being camera and gaussian idcs right?

**Review by vye16 — CHANGES_REQUESTED** [review]
amazing ruilong, I learned so much! I just left a couple of comments about naming and interfaces (talking more offline) and we can merge this in super soon!

**Comment by Linusnie** [comment]
Hi, this looks like a great update! I did some testing for pose gradients and noticed a pretty significant slowdown (roughly 3x). It's probably not noticeable in terms of overall reconstruction time, although might be worth looking into.

There is a workaround by transforming the gaussians to camera space and letting autodiff handle the pose jacobians (as is done in SplaTAM https://github.com/spla-tam/SplaTAM/issues/64). However, this adds some overhead as you need an extra matrix-to-quaternion conversion for rotating the covariance matrices.

My timings experiments can be found here https://github.com/Linusnie/render-splats/blob/main/timings.ipynb

**Comment by liruilong940607** [comment]
> Hi, this looks like a great update! I did some testing for pose gradients and noticed a pretty significant slowdown (roughly 3x). It's probably not noticeable in terms of overall reconstruction time, although might be worth looking into.
> 
> There is a workaround by transforming the gaussians to camera space and letting autodiff handle the pose jacobians (as is done in SplaTAM [spla-tam/SplaTAM#64](https://github.com/spla-tam/SplaTAM/issues/64)). This variant is almost as fast as the legacy pose gradients, but adds some overhead as you need an extra matrix-to-quaternion conversion for rotating the covariance matrices.
> 
> My timings experiments can be found here https://github.com/Linusnie/render-splats/blob/main/timings.ipynb

Hey @Linusnie thanks for this profiling!! It's a very interesting experiment.

First of all, I would like to state that the camera pose gradients are flawed in of `gsplat v.0.1.X`, as the `spherical_harmonics` is differentiable w.r.t view directions, but [the gradients are not implemented](https://github.com/nerfstudio-project/gsplat/blob/main/gsplat/sh.py#L97-L104) (essentially putting a `detach()` there).

It is implemented in this PR to make sure the camera pose received the **full** gradients.

So the slow down you observe should be mainly coming from there, i.e., the d (SH) / d(viewdirs) is currently quite a bottleneck. And by transforming Gaussians into the camera coordinate system (your third experiment), you would actually also need to transform the SH coeffs to the camera coordinate system using the camera pose, which is missing in your currently implementation. So as a result, your third experiment skips the gradient calculation of d (SH) / d(viewdirs), resulting a significant speedup.

But this is an interesting thought! The camera gradients in this PR is numerically correct but we didn't put too much efforts on optimizing it for speed. This idea might be able to give us something if we could figure out how SH would be handled.

**Comment by Linusnie** [comment]
@liruilong940607 I see, thanks for the explanation! Very nice to have the full gradient implemented. I do transform the view directions in the legacy version with this line `view_dirs = bmv(gs_data.rotation.T, torch.nn.functional.normalize(gs_data.means - camera_center))`, which I guess should be more or less equivalent to transforming the SH coefficient (maybe with some difference due to truncation error).

**Comment by liruilong940607** [comment]
@Linusnie Ah that is smart! But I guess the view directions are not transformed in your third experiment (V1.0 transformed Gaussians)? As this would need to be injected into `rasterization` function in V1.0.

**Comment by Linusnie** [comment]
@liruilong940607 that's right. Although I just realized the view directions already depend on the pose in "V1.0 transformed" via the gaussian means (which are transformed by the pose):

```python
camera_center = torch.zeros(3, device='cuda', requires_grad=False)
means = (camera_pose @ gs_data).means
view_dirs = torch.nn.functional.normalize(means - camera_center)
colors = gsplat.cuda._wrapper.spherical_harmonics(gs_data.max_sh_degree, view_dirs, gs_data.features)
torch.autograd.grad(colors.mean(), camera_pose)
# returns nonzero value
```

So looks like the cause would be something else after all. Maybe some overhead due to supporting batched poses?

**Comment by liruilong940607** [comment]
@Linusnie I guess what I was trying to express is that only transforming `view_dirs` is not enough -- i.e., it will end up a different output `colors` from this function (because you are using a camera space view direction to query the SH basis instead of a world space view direction)

```bash
colors = gsplat.cuda._wrapper.spherical_harmonics(gs_data.max_sh_degree, view_dirs, gs_data.features)
```

That being said, the SH coeffs (`gs_data.features`) is actually also coordinate-system-dependent, which should also be "transformed" by the camera pose, so that the above function will give the same output between camera space inputs and world space inputs. But I'm not sure how the SH coeffs can be transformed by camera pose.

**Comment by Linusnie** [comment]
@liruilong940607 I think there should be no need to transform `gs_data.features` directly. For the legacy variant I keep track of the rotation that has been applied to the gaussians via `gs_data.rotation`. This rotation can then be used to transform view directions from camera coordinates to the original SH coordinates. In both cases you actually end up passing the same viewdirs to `spherical_harmonics`: `m - (-R^Tt) = R^T(Rm + t)` (LHS: regular variant, RHS: transformed variant, `m` is the gaussian mean, `R,t` is the pose). So in terms of rendered colors they are equivalent

As you said this rotation would need to be passed as an additional argument to `rasterization` to keep the output correct in V1.0. With my previous comment I just wanted to point out that gradients are propagated through the view directions both with or without this correction, via the gaussian means

**Comment by liruilong940607** [comment]
@Linusnie Ah you are right! I didn't see it through that this transformation leads to the same `viewdirs` as before! But this still wouldn't avoid the computation of d (colors) / d (view_dirs) in `spherical_harmonics`, which I guess meaning it wouldn't speed up things?

```
view_dirs = torch.nn.functional.normalize(means - camera_center) # should additionally be rotated by camera_pose[:3, :3] to be geometrically consistent
colors = gsplat.cuda._wrapper.spherical_harmonics(gs_data.max_sh_degree, view_dirs, gs_data.features)
```

**Comment by Linusnie** [comment]
@liruilong940607 that's the thing, the slow execution for pose gradients seems to rather be connected to whether `viewmats` has requires_grad=True or not. Not on whether gradients are propagated through the spherical harmonics. Based on that `render_splats(camera_pose, gs_data)` is much slower than `render_splats(eye(4), camera_pose @ gs_data)`, even though d(SH)/d(viewdirs) is computed in both cases

**Comment by Linusnie** [comment]
@liruilong940607 I've narrowed it down to this line here https://github.com/nerfstudio-project/gsplat/blob/v1.0/gsplat/rendering.py#L274! Looks like the backwards pass through the re-indexing of `camtoworlds` is what's taking so much time (even though `rindicies` is all zeros). Probably because I'm working with a small example scene (the lego excavator) and one camera.

Using replacing the second part with `camtoworlds[:, :3, 3].repeat((len(cindices), 1))` resolves the issue. This doesn't handle the multi-camera case, but maybe something similar could be made to work.

**Comment by liruilong940607** [comment]
Very interesting! Does that change bring the timing to the same between the two options?

File: gsplat/cuda/csrc/rasterization.cu
Comment (liruilong940607) [reply gsplat/cuda/csrc/rasterization.cu:678]: I'll leave this to next PR ... lol

File: gsplat/cuda/_wrapper.py
Comment (liruilong940607) [reply gsplat/cuda/_wrapper.py:141]: good point! I rename it to `fully_fused_projection`

File: gsplat/cuda/_wrapper.py
Comment (liruilong940607) [reply gsplat/cuda/_wrapper.py]: done! `rindices` -> `camera_ids`, `cindices` -> `gaussian_ids`

File: gsplat/cuda/csrc/rasterization.cu
Comment (liruilong940607) [reply gsplat/cuda/csrc/rasterization.cu:254]: good point. done!

**Review by vye16 — APPROVED** [review]
Nice update on radixsort, 👏 👏

File: CITATION.cff
Comment (maturk) [inline CITATION.cff:1]: Do we update the citation along with this PR or reserve that for later? @liruilong940607 , @vye16

Comment (liruilong940607) [reply CITATION.cff:1]: any reason that we want to reserve it for later?

**Review by maturk — APPROVED** [review]

File: CITATION.cff
Comment (liruilong940607) [reply CITATION.cff:1]: ok communicated offline!

File: gsplat/rendering.py
Comment (liruilong940607) [reply gsplat/rendering.py:289]: Ah good point! that's a bug we should fix on this nerfstudio side!

Comment (liruilong940607) [reply gsplat/rendering.py:289]: Fixed in https://github.com/nerfstudio-project/nerfstudio/pull/3207

**Pull request merged**
